### PR TITLE
Align code base to v0.11

### LINKED ISSUE #5119: Align to spec v0.11
With the release of [v0.11](https://github.com/ethereum/eth2.0-specs/releases/tag/v0.11.0) to facilitate new shading proposal. This PR tracks to the progress to align Prysm to v.0.11 which is important for multi client test net. The experimental branch for this is: 

The v.0.9 consists of the following:
https://github.com/prysmaticlabs/prysm/tree/spec-v0.11

## Beacon chain
- [x] Adjust hysteresis to avoid initial over-deposit incentive (https://github.com/ethereum/eth2.0-specs/pull/1627)
- [x] Add proposer_index to BeaconBlock (https://github.com/ethereum/eth2.0-specs/pull/1626)
- [x] Handle rewards overflow (https://github.com/ethereum/eth2.0-specs/pull/1635)
- [x] Add genesis_validators_root to BeaconState to support better fork separation (https://github.com/ethereum/eth2.0-specs/pull/1614, https://github.com/ethereum/eth2.0-specs/pull/1652)
- [x] Denominate Eth1 voting period in epochs (https://github.com/ethereum/eth2.0-specs/pull/1649)
- [x] Attestation must match target to match head for rewards (https://github.com/ethereum/eth2.0-specs/pull/1653)

##  Fork choice
- [x] Pull out some helper functions for better code reuse across forks [No Substantive Change] (https://github.com/ethereum/eth2.0-specs/pull/1504)

##  Validator

- [x] BeaconBlock slashing is per-slot (reflect this in v-guide) (https://github.com/ethereum/eth2.0-specs/pull/1612)

##  Networking
- [x] Add libp2p-noise specification (https://github.com/ethereum/eth2.0-specs/pull/1607)
- [x] Remove head_block_root from BeaconBlocksByRange (https://github.com/ethereum/eth2.0-specs/pull/1604)
- [x] Add subnet validations for DoS resistance (https://github.com/ethereum/eth2.0-specs/pull/1615) #5158
- [x] Add lower bound slot condition on block gossip (https://github.com/ethereum/eth2.0-specs/pull/1616) #5166 
- [x] Add DoS prevention validation conditions to additional gossipsub topics (https://github.com/ethereum/eth2.0-specs/pull/1617) 
- [x] Use snappy frames to support streaming (https://github.com/ethereum/eth2.0-specs/pull/1606)
- [x] Add eth2 key/value to ENR (https://github.com/ethereum/eth2.0-specs/pull/1614)
- [x] Add fork digest to gossipsub topics for simple fork versioning of messages and to Status message (https://github.com/ethereum/eth2.0-specs/pull/1652)

### PR DESCRIPTION
**Pull request by terencechain**
Builds on #4804 and incorporates all changes.

Resolves #5119

### CODE REVIEW

**Comment by terencechain** [comment]
Working on `Adjust hysteresis to avoid initial over-deposit incentive`

File: beacon-chain/core/state/benchmarks_test.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 var runAmount = 25
 
 func TestBenchmarkExecuteStateTransition(t *testing.T) {
+	t.Skip("To be resolved until 5119 gets in")
```
Comment (rauljordan) [inline beacon-chain/core/state/benchmarks_test.go:20]: We need to unskip this

File: proto/testing/ssz_static_test.go [hunk 2]
```diff
@@ -130,8 +130,13 @@ func UnmarshalledSSZ(t *testing.T, serializedBytes []byte, folderName string) (i
 		return nil, nil
 	case "Eth1Data":
 		obj = &ethpb.Eth1Data{}
+	case "Eth1Block":
+		t.Skip("Unused type")
```
Comment (rauljordan) [inline proto/testing/ssz_static_test.go:134]: Delete?

File: shared/bls/spectest/aggregate_verify_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+				}
+				t.Fatalf("Cannot unmarshal input to signature: %v", err)
+			}
+
+			verified := sig.AggregateVerify(pubkeys, msgs)
+			if verified != test.Output {
+				t.Fatalf("Signature does not match the expected verification output. "+
+					"Expected %#v but received %#v for test case %d", test.Output, verified, i)
+			}
+			t.Log("Success")
```
Comment (rauljordan) [inline shared/bls/spectest/aggregate_verify_test.go:66]:
```suggestion
```

File: shared/bls/spectest/verify_test.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+				}
+				t.Fatalf("Cannot unmarshal input to signature: %v", err)
+			}
+
+			verified := sig.Verify(msgBytes, pk)
+			if verified != test.Output {
+				t.Fatalf("Signature does not match the expected verification output. "+
+					"Expected %#v but received %#v for test case %d", test.Output, verified, i)
+			}
+			t.Log("Success")
```
Comment (rauljordan) [inline shared/bls/spectest/verify_test.go:58]:
```suggestion
```

File: shared/testutil/deposits_test.go [hunk 5]
```diff
@@ -8,7 +8,9 @@ import (
 )
 
 func TestSetupInitialDeposits_1024Entries(t *testing.T) {
+	t.Skip("To be resolved until 5119 gets in")
```
Comment (rauljordan) [inline shared/testutil/deposits_test.go:11]: Need to unskip

File: beacon-chain/sync/subscriber.go [hunk 6]
```diff
@@ -316,11 +292,93 @@ func (r *Service) subscribeDynamic(topicFormat string, determineSubsLen func() i
...
+// find if we have peers who are subscribed to the same subnet
+func (r *Service) validPeersExist(subnetTopic string, idx uint64) bool {
+	numOfPeers := r.p2p.PubSub().ListPeers(subnetTopic + r.p2p.Encoding().ProtocolSuffix())
+	return len(r.p2p.Peers().SubscribedToSubnet(idx)) > 0 || len(numOfPeers) > 0
+}
+
+// Add fork digest to topic.
+func (r *Service) addDigestToTopic(topic string) string {
+	if !strings.Contains(topic, "%x") {
+		log.Fatal("Topic does not have appropriate formatter for digest")
```
Comment (rauljordan) [inline beacon-chain/sync/subscriber.go:377]: Maybe we should say what the topic formatter should be

File: beacon-chain/archiver/service_test.go [hunk 7]
```diff
@@ -433,18 +424,12 @@ func setupState(validatorCount uint64) (*stateTrie.BeaconState, error) {
 
 	// We initialize a head state that has attestations from participated
 	// validators in a simulated fashion.
-	return stateTrie.InitializeFromProto(&pb.BeaconState{
-		Slot:                       (2 * params.BeaconConfig().SlotsPerEpoch) - 1,
-		Validators:                 validators,
-		Balances:                   balances,
-		BlockRoots:                 make([][]byte, params.BeaconConfig().SlotsPerHistoricalRoot),
-		Slashings:                  []uint64{0, 1e9, 1e9},
```
Comment (shayzluf) [inline beacon-chain/archiver/service_test.go:441]: missing st.SetSlashings

File: tools/faucet/server.go [hunk 8]
```diff
@@ -24,7 +24,7 @@ import (
 
 const ipLimit = 5
 
-var fundingAmount = big.NewInt(3.5 * params.Ether)
+var fundingAmount = big.NewInt(32.5 * 1e9)
```
Comment (prestonvanloon) [inline tools/faucet/server.go:27]:
```suggestion
var fundingAmount = big.NewInt(32.5 * params.Ether)
```

File: beacon-chain/core/blocks/block_operations_fuzz_test.go [hunk 9]
```diff
@@ -44,31 +43,6 @@ func TestFuzzProcessBlockHeader_10000(t *testing.T) {
 	}
 }
 
-func TestFuzzverifySigningRoot_10000(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/core/blocks/block_operations_fuzz_test.go:47]: Nvm, I see it was moved somewhere else

File: beacon-chain/blockchain/head.go [hunk 10]
```diff
@@ -59,8 +59,8 @@ func (s *Service) saveHead(ctx context.Context, headRoot [32]byte) error {
 
 	// If the head state is not available, just return nil.
 	// There's nothing to cache
-	if featureconfig.Get().NewStateMgmt {
-		if !s.stateGen.StateSummaryExists(ctx, headRoot) {
+	if !featureconfig.Get().DisableNewStateMgmt {
+		if !s.stateGen.StateSummaryExists(ctx, headRoot) && !s.beaconDB.HasState(ctx, headRoot) {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/blockchain/head.go:63]: Can remove `!s.beaconDB.HasState(ctx, headRoot)`, will fix the godoc. A better phase:
`StateSummaryExists returns true if the corresponding state of the input block either exists in the DB or state summary can be generated by state gen`

File: beacon-chain/operations/attestations/kv/unaggregated.go [hunk 11]
```diff
@@ -52,6 +52,22 @@ func (p *AttCaches) UnaggregatedAttestations() []*ethpb.Attestation {
 	return atts
 }
 
+// UnAggregatedAttestationsBySlotIndex returns the unaggregated attestations in cache,
+// filtered by committee index and slot.
+func (p *AttCaches) UnAggregatedAttestationsBySlotIndex(slot uint64, committeeIndex uint64) []*ethpb.Attestation {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/operations/attestations/kv/unaggregated.go:57]: Agree, we should delete these

File: beacon-chain/p2p/monitoring.go [hunk 12]
```diff
@@ -6,11 +6,6 @@ import (
 )
 
 var (
-	p2pTopicPeerCount = promauto.NewGaugeVec(prometheus.GaugeOpts{
-		Name: "p2p_topic_peer_count",
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/p2p/monitoring.go:10]: @rauljordan @nisdas

File: beacon-chain/p2p/options.go [hunk 13]
```diff
@@ -42,8 +41,8 @@ func buildOptions(cfg *Config, ip net.IP, priKey *ecdsa.PrivateKey) []libp2p.Opt
 		options = append(options, libp2p.AddrsFactory(withRelayAddrs(cfg.RelayNodeAddr)))
 	}
 	if cfg.HostAddress != "" {
-		options = append(options, libp2p.AddrsFactory(func(addrs []multiaddr.Multiaddr) []multiaddr.Multiaddr {
-			external, err := multiAddressBuilder(cfg.HostAddress, cfg.TCPPort)
+		options = append(options, libp2p.AddrsFactory(func(addrs []ma.Multiaddr) []ma.Multiaddr {
+			external, err := ma.NewMultiaddr(fmt.Sprintf("/ip4/%s/tcp/%d", cfg.HostAddress, cfg.TCPPort))
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/p2p/options.go:45]: @rauljordan @nisdas

File: WORKSPACE [hunk 14]
```diff
@@ -285,6 +282,13 @@ http_archive(
     urls = ["https://github.com/edenhill/librdkafka/archive/v1.2.1.tar.gz"],
 )
 
+http_archive(
+    name = "com_github_herumi_bls_eth_go_binary",
+    sha256 = "3b155ff597c307b5b0875c3b1ffc3beaa0a6634ee1cfad8768041b61c47eac39",
+    strip_prefix = "bls-eth-go-binary-946ee977cd0f585757741dda65073cbd84a160ac",
+    url = "https://github.com/prysmaticlabs/bls-eth-go-binary/archive/946ee977cd0f585757741dda65073cbd84a160ac.zip",
```
[the remark this answers is not in the document]
Comment (terencechain) [reply WORKSPACE:292]: @nisdas

File: beacon-chain/p2p/monitoring.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/p2p/monitoring.go:10]: This was moved. I saw it later in the PR. Resolving

File: beacon-chain/p2p/service.go [hunk 15]
```diff
@@ -148,6 +175,21 @@ func (s *Service) Start() {
 		return
 	}
 
+	// Waits until the state is initialized via an event feed.
+	// Check if we have a genesis time / genesis state
+	// used for fork-related data when connecting peers.
+	genesisState, err := s.beaconDB.GenesisState(s.ctx)
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/p2p/service.go:181]: @rauljordan

File: beacon-chain/sync/decode_pubsub.go [hunk 16]
```diff
@@ -26,3 +27,10 @@ func (r *Service) decodePubsubMessage(msg *pubsub.Message) (proto.Message, error
 	}
 	return m, nil
 }
+
+// Replaces our fork digest with the formatter.
+func (r *Service) replaceForkDigest(topic string) string {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/sync/decode_pubsub.go:32]: @nisdas

File: beacon-chain/core/helpers/committee.go [hunk 17]
```diff
@@ -235,85 +240,7 @@ func CommitteeAssignments(state *stateTrie.BeaconState, epoch uint64) (map[uint6
...
-//    assert epoch <= next_epoch
-//
-//    start_slot = compute_start_slot_at_epoch(epoch)
-//    for slot in range(start_slot, start_slot + SLOTS_PER_EPOCH):
-//        for index in range(get_committee_count_at_slot(state, Slot(slot))):
-//            committee = get_beacon_committee(state, Slot(slot), CommitteeIndex(index))
-//            if validator_index in committee:
-//                return committee, CommitteeIndex(index), Slot(slot)
-//    return None
-func CommitteeAssignment(
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/helpers/committee.go:271]: @terencechain

File: beacon-chain/core/helpers/committee_test.go [hunk 18]
```diff
@@ -191,148 +190,6 @@ func TestVerifyBitfieldLength_OK(t *testing.T) {
...
-	index := uint64(10000)
-	_, _, _, _, err := CommitteeAssignment(state, 1, index)
-	if err != nil && !strings.Contains(err.Error(), "not found in assignments") {
-		t.Errorf("Wanted 'not found in assignments', received %v", err)
-	}
-}
-
-// Test helpers.CommitteeAssignments against the results of helpers.CommitteeAssignment by validator
-// index. Warning: this test is a bit slow!
-func TestCommitteeAssignments_AgreesWithSpecDefinitionMethod(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/helpers/committee_test.go:294]: @terencechain

File: beacon-chain/core/helpers/slot_epoch.go [hunk 19]
```diff
@@ -89,15 +89,17 @@ func SlotsSinceEpochStarts(slot uint64) uint64 {
 	return slot - StartSlot(SlotToEpoch(slot))
 }
 
-// Allow for slots "from the future" within a certain tolerance.
-const timeShiftTolerance = 10 // ms
+// TimeShiftTolerance specifies the tolerance threshold for slots "from the future".
+const TimeShiftTolerance = 500 * time.Millisecond // ms
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/core/helpers/slot_epoch.go:93]: @rauljordan

File: beacon-chain/operations/attestations/kv/aggregated.go [hunk 20]
```diff
@@ -3,10 +3,68 @@ package kv
...
+			}
+		}
+		if err := p.SaveAggregatedAttestations(aggregatedAtts); err != nil {
+			return err
+		}
+	}
+
+	// Remove the unaggregated attestations from the pool that were successfully aggregated.
+	for _, att := range unaggregatedAtts {
+		h, err := ssz.HashTreeRoot(att)
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/operations/attestations/kv/aggregated.go:53]: @terencechain

File: beacon-chain/operations/attestations/kv/unaggregated.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/operations/attestations/kv/unaggregated.go:57]: @terencechain

File: beacon-chain/operations/attestations/kv/unaggregated.go [hunk 21]
```diff
@@ -73,6 +89,29 @@ func (p *AttCaches) DeleteUnaggregatedAttestation(att *ethpb.Attestation) error
 	return nil
 }
 
+// DeleteUnaggregatedAttestations deletes the unaggregated attestations in cache.
+func (p *AttCaches) DeleteUnaggregatedAttestations(atts []*ethpb.Attestation) error {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/operations/attestations/kv/unaggregated.go:93]: @terencechain

File: beacon-chain/rpc/validator/aggregator.go [hunk 22]
```diff
@@ -56,30 +57,47 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *ethpb.Aggreg
...
+		return nil, status.Error(codes.InvalidArgument, "Signed aggregate request can't be nil")
 	}
 
-	return &ethpb.AggregationResponse{}, nil
+	if err := as.P2P.Broadcast(ctx, req.SignedAggregateAndProof); err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not broadcast signed aggregated attestation: %v", err)
+	}
+
+	log.WithFields(logrus.Fields{
+		"slot":            req.SignedAggregateAndProof.Message.Aggregate.Data.Slot,
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/rpc/validator/aggregator.go:96]: @terencechain

File: beacon-chain/rpc/validator/aggregator.go [hunk 23]
```diff
@@ -56,30 +57,47 @@ func (as *Server) SubmitAggregateAndProof(ctx context.Context, req *ethpb.Aggreg
...
+		SelectionProof:  req.SlotSignature,
+		AggregatorIndex: validatorIndex,
+	}
+	return &ethpb.AggregateSelectionResponse{AggregateAndProof: a}, nil
+}
+
+// SubmitSignedAggregateSelectionProof is called by a validator to broadcast a signed
+// aggregated and proof object.
+func (as *Server) SubmitSignedAggregateSelectionProof(ctx context.Context, req *ethpb.SignedAggregateSubmitRequest) (*ethpb.SignedAggregateSubmitResponse, error) {
+	if req.SignedAggregateAndProof == nil {
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/rpc/validator/aggregator.go:87]: @terencechain

File: proto/beacon/rpc/v1/BUILD.bazel [hunk 24]
```diff
@@ -0,0 +1,14 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
+
+go_library(
+    name = "go_default_library",
+    srcs = ["slasher.pb.go"],
```
[the remark this answers is not in the document]
Comment (terencechain) [reply proto/beacon/rpc/v1/BUILD.bazel:5]: @0xKiwi @shayzluf

File: proto/beacon/rpc/v1/slasher.pb.go [hunk 25]
```diff
@@ -0,0 +1,2042 @@
+// Code generated by protoc-gen-gogo. DO NOT EDIT.
+// source: proto/beacon/rpc/v1/slasher.proto
```
[the remark this answers is not in the document]
Comment (terencechain) [reply proto/beacon/rpc/v1/slasher.pb.go:2]: @0xKiwi @shayzluf

File: validator/node/node.go [hunk 26]
```diff
@@ -219,35 +219,25 @@ func selectKeyManager(ctx *cli.Context) (keymanager.KeyManager, error) {
...
-				log.Warn("Implicit selection of keymanager is deprecated.  Please use --keymanager=keystore or select a different keymanager")
-			} else {
-				opts = fmt.Sprintf(`{"passphrase":%q}`, passphrase)
-				log.Warn(`Implicit selection of keymanager is deprecated.  Please use --keymanager=keystore --keymanageropts='{"passphrase":"<password>"}' or select a different keymanager`)
-			}
 		}
 	}
 
+	if manager == "" {
+		return nil, fmt.Errorf("please supply a keymanager with --keymanager")
```
[the remark this answers is not in the document]
Comment (terencechain) [reply validator/node/node.go:226]: @rauljordan

File: beacon-chain/archiver/service_test.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/archiver/service_test.go:441]: @terencechain

File: WORKSPACE (hunk 14 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply WORKSPACE:292]: This shouldn't go here anyway. Going to delete it and update the appropriate file.

File: beacon-chain/archiver/service_test.go (hunk 7 above)
Comment (rauljordan) [reply beacon-chain/archiver/service_test.go:441]: No need, it is already done by testutil.NewBeaconState no? @terencechain

File: beacon-chain/core/helpers/committee_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/core/helpers/committee_test.go:294]: It is no longer relevant @prestonvanloon, this function is checking for compatibility with a deprecated function that has been deleted from the codebase. Our new committee assignments is the spec definition. I believe this test had poor naming, it was built during a transition period of v0.8 to spec v0.9 a long time ago.

File: validator/usage.go [hunk 27]
```diff
@@ -72,13 +72,10 @@ var appHelpFlagGroups = []flagGroup{
 	{
 		Name: "validator",
 		Flags: []cli.Flag{
-			flags.NoCustomConfigFlag,
 			flags.BeaconRPCProviderFlag,
 			flags.CertFlag,
 			flags.KeyManager,
 			flags.KeyManagerOpts,
-			flags.KeystorePathFlag,
-			flags.PasswordFlag,
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply validator/usage.go:81]: need to add the concept of deprecated flags for the validator

File: beacon-chain/core/helpers/slot_epoch.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/core/helpers/slot_epoch.go:93]: Can be flexible, I'll bring it down to 10ms

File: beacon-chain/p2p/service.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/p2p/service.go:181]: Correct, all we need is to listen to some feed from somewhere that the state was initialized

File: beacon-chain/sync/subscriber_handlers.go [hunk 28]
```diff
@@ -49,6 +58,8 @@ func (r *Service) proposerSlashingSubscriber(ctx context.Context, msg proto.Mess
 	}
 	// Do some nil checks to prevent easy DoS'ing of this handler.
 	if ps.Header_1 != nil && ps.Header_1.Header != nil {
+		r.setProposerSlashingIndexSeen(ps.Header_1.Header.ProposerIndex)
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/sync/subscriber_handlers.go:61]: @rauljordan

File: beacon-chain/sync/subscriber_handlers.go [hunk 29]
```diff
@@ -30,6 +37,8 @@ func (r *Service) attesterSlashingSubscriber(ctx context.Context, msg proto.Mess
 	}
 	// Do some nil checks to prevent easy DoS'ing of this handler.
 	if as != nil && as.Attestation_1 != nil && as.Attestation_1.Data != nil {
+		r.setAttesterSlashingIndicesSeen(as.Attestation_1.AttestingIndices, as.Attestation_2.AttestingIndices)
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply beacon-chain/sync/subscriber_handlers.go:40]: @rauljordan

File: beacon-chain/archiver/service_test.go (hunk 7 above)
Comment (terencechain) [reply beacon-chain/archiver/service_test.go:441]: I think so

File: beacon-chain/core/helpers/slot_epoch.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/core/helpers/slot_epoch.go:93]: This is defined in the spec, which is why it was changed

[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/core/helpers/slot_epoch.go:93]: You can find it here https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/p2p-interface.md#configuration , under 
```
MAXIMUM_GOSSIP_CLOCK_DISPARITY
```

File: shared/interop/generate_genesis_state.go [hunk 30]
```diff
@@ -149,8 +150,15 @@ func createDepositData(privKey *bls.SecretKey, pubKey *bls.PublicKey) (*ethpb.De
 	if err != nil {
 		return nil, err
 	}
-	domain := bls.ComputeDomain(params.BeaconConfig().DomainDeposit)
-	di.Signature = privKey.Sign(sr[:], domain).Marshal()
+	domain, err := helpers.ComputeDomain(params.BeaconConfig().DomainDeposit, nil, nil)
```
[the remark this answers is not in the document]
Comment (nisdas) [reply shared/interop/generate_genesis_state.go:153]: why ?

File: beacon-chain/db/kv/check_historical_state.go [hunk 31]
```diff
@@ -32,17 +32,17 @@ func (kv *Store) ensureNewStateServiceCompatible(ctx context.Context) error {
 	regenHistoricalStatesConfirmed := false
 	var err error
 	if historicalStateDeleted {
-		actionText := "Looks like you stopped using --new-state-mgmt. To reuse it, the node will need " +
-			"to generate and save historical states. The process may take a while, - do you want to proceed? (Y/N)"
-		deniedText := "Historical states will not be generated. Please remove usage --new-state-mgmt"
+		actionText := "--disable-new-state-mgmt was used. To proceed without the flag, the db will need " +
+			"to generate and save historical states. This process may take a while, - do you want to proceed? (Y/N)"
+		deniedText := "Historical states will not be generated. Please continue use --disable-new-state-mgmt"
```
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/db/kv/check_historical_state.go:37]: Huh? The PR says `Please continue use --disable-new-state-mgmt`

File: shared/keystore/deposit_input_test.go [hunk 32]
```diff
@@ -5,13 +5,14 @@ import (
...
 	"github.com/prysmaticlabs/go-ssz"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
 	"github.com/prysmaticlabs/prysm/shared/bls"
-	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	"github.com/prysmaticlabs/prysm/shared/keystore"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
 func TestDepositInput_GeneratesPb(t *testing.T) {
+	t.Skip("To be resolved until 5119 gets in")
```
[the remark this answers is not in the document]
Comment (terencechain) [reply shared/keystore/deposit_input_test.go:15]: No, please unskip it

File: validator/main.go [hunk 33]
```diff
@@ -82,68 +77,6 @@ func main() {
...
 	app.Version = version.GetVersion()
 	app.Action = startNode
-	app.Commands = []*cli.Command{
-		{
-			Name:     "accounts",
-			Category: "accounts",
-			Usage:    "defines useful functions for interacting with the validator client's account",
-			Subcommands: []*cli.Command{
-				{
-					Name: "create",
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply validator/main.go:92]: This change is being reverted in #5416

File: third_party/herumi/herumi.bzl [hunk 34]
```diff
@@ -8,21 +8,15 @@ Herumi's BLS library for go depends on
...
         urls = [
-            "https://github.com/herumi/bls-eth-go-binary/archive/147ed25f233ed0b211e711ed8271606540c58064.tar.gz",
+            "https://github.com/herumi/bls-eth-go-binary/archive/da18d415993a059052dfed16711f2b3bd03c34b8.tar.gz",
         ],
-        sha256 = "bbd04f3354f12982e4ef32c62eb13ceb183303ada1ee69e2869553ed35134321",
+        sha256 = "69080ca634f8aaeb0950e19db218811f4bb920a054232e147669ea574ba11ef0",
         build_file = "@prysm//third_party/herumi:bls_eth_go_binary.BUILD",
-        # TODO(4804): Delete this patch after updating this archive to commit 381c62473c28af84f424cfb1521c97e48289174a or later.
-        patches = [
-            "@prysm//third_party/herumi:bls_eth_go_binary_serialization_alloc_fix.patch",  # Integrates changes from PR #5.
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply third_party/herumi/herumi.bzl:23]: This is being handled in #5414

File: beacon-chain/db/kv/check_historical_state.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (0xKiwi) [reply beacon-chain/db/kv/check_historical_state.go:37]: The english just seems weird, its on the denied text for a confirmation of the usage of the flag. "Please continue use --disable" when you deny the flag? Maybe its supposed to be "Please continue using --disable"?

File: shared/keystore/deposit_input_test.go (hunk 32 above)
[the remark this answers is not in the document]
Comment (0xKiwi) [reply shared/keystore/deposit_input_test.go:15]: Kk, on it

File: beacon-chain/db/kv/check_historical_state.go (hunk 31 above)
[the remark this answers is not in the document]
Comment (terencechain) [reply beacon-chain/db/kv/check_historical_state.go:37]: sure we can update it

File: beacon-chain/core/helpers/committee_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/core/helpers/committee_test.go:294]: Oh did they rewrite the committee assignments? It was extremely extremely slow O(n^2) so we wrote our own which was O(n) and we maintained this test to ensure we were compatible with the tested spec definition

File: beacon-chain/core/epoch/precompute/reward_penalty.go [hunk 35]
```diff
@@ -93,14 +95,18 @@ func attestationDelta(state *stateTrie.BeaconState, bp *Balance, v *Validator) (
 
 	// Process target reward / penalty
 	if v.IsPrevEpochTargetAttester && !v.IsSlashed {
-		r += br * bp.PrevEpochTargetAttesters / bp.CurrentEpoch
+		inc := params.BeaconConfig().EffectiveBalanceIncrement
+		rewardNumerator := br * bp.PrevEpochAttesters / inc
```
Comment (protolambda) [inline beacon-chain/core/epoch/precompute/reward_penalty.go:99]: This is a critical bug :beetle: :bug:

File: beacon-chain/core/epoch/precompute/reward_penalty.go [hunk 36]
```diff
@@ -93,14 +95,18 @@ func attestationDelta(state *stateTrie.BeaconState, bp *Balance, v *Validator) (
...
+		r += rewardNumerator / (bp.CurrentEpoch / inc)
 	} else {
 		p += br
 	}
 
 	// Process head reward / penalty
 	if v.IsPrevEpochHeadAttester && !v.IsSlashed {
-		r += br * bp.PrevEpochHeadAttesters / bp.CurrentEpoch
+		inc := params.BeaconConfig().EffectiveBalanceIncrement
+		rewardNumerator := br * bp.PrevEpochAttesters / inc
```
Comment (protolambda) [inline beacon-chain/core/epoch/precompute/reward_penalty.go:108]: This is a critical bug as well :beetle: :bug:

**Pull request merged**